File: 단계별로 풀어보기/Step 16/12865.cpp
#include <iostream>
using namespace std;

int weight[101];
int value[101];
int dp[101][100001];

int main() {
    int N, K;
    int i, j;

    cin >> N >> K;
    for(i=1; i<=N; i++) {
        cin >> weight[i] >> value[i];
    }

    for(i=1; i<=N; i++) {
        for(j=1; j<=K; j++) {
            dp[i][j] = dp[i-1][j];
            if(j-weight[i] >= 0)
                dp[i][j] = max(dp[i][j], dp[i-1][j-weight[i]]+value[i]);
        }
    }

    cout << dp[N][K] << endl;
    return 0;
}

File: StepByStep/Python/Step_16_동적 계획법 1/12865.py
N, K = map(int, input().split())
weight = [0] * (N+1)
value = [0] * (N+1)
for i in range(1, N+1):
    weight[i], value[i] = map(int, input().split())

dp = [[0] * (K+1) for _ in range(N+1)]
for j in range(1, K+1):
    for i in range(1, N+1):
        if weight[i] > j:
            dp[i][j] = dp[i-1][j]
            continue
        dp[i][j] = max(dp[i-1][j], value[i]+dp[i-1][j-weight[i]])

print(dp[N][K])